### PR TITLE
Add C++ Client Encryptor implementation

### PR DESCRIPTION
**Pull request by ipetr0v**
This PR adds an implementation for the Client Encryptor as part of C++ Hybrid Encryption implementation.

Ref https://github.com/project-oak/oak/issues/4026

### CODE REVIEW

**Comment by ipetr0v** [comment]
Curently blocked, because we need to update Protobuf version, to make it work with `absl::string_view`.

File: cc/crypto/client_encryptor.cc [hunk 1]
```diff
@@ -17,7 +17,73 @@
...
+
+absl::StatusOr<std::unique_ptr<ClientEncryptor>> ClientEncryptor::Create(
+    absl::string_view serialized_server_public_key) {
+  absl::StatusOr<SenderContext> setup_base_sender_result =
+      SetupBaseSender(serialized_server_public_key, OAK_HPKE_INFO);
+  if (!setup_base_sender_result.ok()) {
+    return setup_base_sender_result.status();
+  }
+  SenderContext sender_context = std::move(*setup_base_sender_result);
+  return std::make_unique<ClientEncryptor>(ClientEncryptor(sender_context));
```
Comment (pmcgrath17) [inline cc/crypto/client_encryptor.cc:39]: keep using setup_base_sender_result variable. The absl::StatusOr class is similar to pointers. If you want to access the object (assuming status is ok()), you can get the object using *

File: cc/crypto/client_encryptor.cc [hunk 2]
```diff
@@ -17,7 +17,73 @@
...
+
+absl::StatusOr<std::string> ClientEncryptor::Encrypt(absl::string_view plaintext,
+                                                     absl::string_view associated_data) {
+  // Encrypt request.
+  absl::StatusOr<std::string> seal_result =
+      sender_request_context_->Seal(plaintext, associated_data);
+  if (!seal_result.ok()) {
+    return seal_result.status();
+  }
+  std::string ciphertext = std::move(*seal_result);
```
Comment (pmcgrath17) [inline cc/crypto/client_encryptor.cc:50]: same as above comment

File: cc/crypto/client_encryptor.cc [hunk 3]
```diff
@@ -17,7 +17,73 @@
...
+    return absl::InvalidArgumentError("couldn't deserialize response");
+  }
+
+  // Decrypt response.
+  absl::StatusOr<std::string> open_result = sender_response_context_->Open(
+      response.encrypted_message().ciphertext(), response.encrypted_message().associated_data());
+  if (!open_result.ok()) {
+    return open_result.status();
+  }
+  std::string plaintext = std::move(*open_result);
```
Comment (pmcgrath17) [inline cc/crypto/client_encryptor.cc:84]: same as above. Feel free to change 'open_result' variable name to 'plaintext'

File: cc/crypto/client_encryptor.cc [hunk 4]
```diff
@@ -17,7 +17,73 @@
...
+  // Create request message.
+  EncryptedRequest request;
+  *request.mutable_encrypted_message()->mutable_ciphertext() = ciphertext;
+  *request.mutable_encrypted_message()->mutable_associated_data() = associated_data;
+
+  // Encapsulated public key is only sent in the initial request message of the session.
+  if (!serialized_encapsulated_public_key_.empty()) {
+    *request.mutable_serialized_encapsulated_public_key() = serialized_encapsulated_public_key_;
+    serialized_encapsulated_public_key_ = "";
+  }
```
Comment (pmcgrath17) [inline cc/crypto/client_encryptor.cc:61]: This seems a little sketchy... What if the class has a static boolean variable to determine if it has sent the encap public key already?

**Review by pmcgrath17 — COMMENTED** [review]
Looks good!

File: cc/crypto/client_encryptor.cc (hunk 1 above)
Comment (ipetr0v) [reply cc/crypto/client_encryptor.cc]: Done

File: cc/crypto/client_encryptor.cc (hunk 2 above)
Comment (ipetr0v) [reply cc/crypto/client_encryptor.cc]: Done

File: cc/crypto/client_encryptor.cc (hunk 3 above)
Comment (ipetr0v) [reply cc/crypto/client_encryptor.cc]: Done

**Review by pmcgrath17 — APPROVED** [review]
Thanks

**Pull request merged**